File: bridge-integration/sdk/python/bridge_sdk.py
"""
Bridge API SDK for Python
Complete integration with Bridge payment platform
"""
import os
from typing import Dict, List, Optional, Any
from enum import Enum
import httpx
from pydantic import BaseModel, Field


class BridgeEnvironment(str, Enum):
    SANDBOX = "sandbox"
    PRODUCTION = "production"


class KYCStatus(str, Enum):
    NOT_STARTED = "not_started"
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"


class WalletType(str, Enum):
    CRYPTO = "crypto"
    LIQUIDITY = "liquidity"


class TransferStatus(str, Enum):
    PENDING = "pending"
    COMPLETED = "completed"
    FAILED = "failed"


class Customer(BaseModel):
    id: str
    email: str
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    kyc_status: KYCStatus = KYCStatus.NOT_STARTED
    created_at: str
    metadata: Dict[str, Any] = Field(default_factory=dict)


class Wallet(BaseModel):
    id: str
    customer_id: str
    wallet_type: WalletType
    currency: str
    address: str
    balance: float = 0.0
    created_at: str


class Transfer(BaseModel):
    id: str
    from_wallet_id: str
    to_wallet_id: str
    amount: float
    currency: str
    status: TransferStatus
    created_at: str
    completed_at: Optional[str] = None


class Card(BaseModel):
    id: str
    customer_id: str
    last_four: str
    brand: str
    status: str
    created_at: str


class BridgeAPIError(Exception):
    """Base exception for Bridge API errors"""
    pass


class BridgeSDK:
    """Complete Bridge API SDK with comprehensive functionality"""
    
    def __init__(
        self,
        api_key: str,
        environment: BridgeEnvironment = BridgeEnvironment.SANDBOX,
        timeout: int = 30
    ):
        self.api_key = api_key
        self.environment = environment
        self.timeout = timeout
        self.base_url = self._get_base_url()
        self.client = httpx.Client(
            headers=self._get_headers(),
            timeout=timeout
        )
    
    def _get_base_url(self) -> str:
        """Get API base URL based on environment"""
        if self.environment == BridgeEnvironment.SANDBOX:
            return "https://api.sandbox.bridge.xyz/v1"
        return "https://api.bridge.xyz/v1"
    
    def _get_headers(self) -> Dict[str, str]:
        """Get default headers for API requests"""
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "User-Agent": "BridgeSDK-Python/1.0"
        }
    
    def _handle_response(self, response: httpx.Response) -> Dict[str, Any]:
        """Handle API response and errors"""
        try:
            response.raise_for_status()
            return response.json()
        except httpx.HTTPStatusError as e:
            error_msg = f"Bridge API error: {e.response.status_code}"
            try:
                error_data = e.response.json()
                error_msg += f" - {error_data.get('message', '')}"
            except Exception:
                pass
            raise BridgeAPIError(error_msg) from e
    
    # Customer Management
    def create_customer(
        self,
        email: str,
        first_name: Optional[str] = None,
        last_name: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> Customer:
        """Create a new customer"""
        payload = {
            "email": email,
            "first_name": first_name,
            "last_name": last_name,
            "metadata": metadata or {}
        }
        response = self.client.post(f"{self.base_url}/customers", json=payload)
        data = self._handle_response(response)
        return Customer(**data)
    
    def get_customer(self, customer_id: str) -> Customer:
        """Get customer details"""
        response = self.client.get(f"{self.base_url}/customers/{customer_id}")
        data = self._handle_response(response)
        return Customer(**data)
    
    def list_customers(self, limit: int = 100, offset: int = 0) -> List[Customer]:
        """List all customers"""
        params = {"limit": limit, "offset": offset}
        response = self.client.get(f"{self.base_url}/customers", params=params)
        data = self._handle_response(response)
        return [Customer(**customer) for customer in data.get("customers", [])]
    
    def delete_customer(self, customer_id: str) -> bool:
        """Delete a customer"""
        response = self.client.delete(f"{self.base_url}/customers/{customer_id}")
        self._handle_response(response)
        return True
    
    def start_kyc(self, customer_id: str) -> Dict[str, Any]:
        """Start KYC verification process"""
        response = self.client.post(f"{self.base_url}/customers/{customer_id}/kyc")
        return self._handle_response(response)
    
    # Wallet Management
    def create_wallet(
        self,
        customer_id: str,
        currency: str,
        wallet_type: WalletType = WalletType.CRYPTO
    ) -> Wallet:
        """Create a new wallet"""
        payload = {
            "customer_id": customer_id,
            "currency": currency,
            "wallet_type": wallet_type.value
        }
        response = self.client.post(f"{self.base_url}/wallets", json=payload)
        data = self._handle_response(response)
        return Wallet(**data)
    
    def get_wallet(self, wallet_id: str) -> Wallet:
        """Get wallet details"""
        response = self.client.get(f"{self.base_url}/wallets/{wallet_id}")
        data = self._handle_response(response)
        return Wallet(**data)
    
    def list_wallets(self, customer_id: str) -> List[Wallet]:
        """List all wallets for a customer"""
        params = {"customer_id": customer_id}
        response = self.client.get(f"{self.base_url}/wallets", params=params)
        data = self._handle_response(response)
        return [Wallet(**wallet) for wallet in data.get("wallets", [])]
    
    def get_wallet_balance(self, wallet_id: str) -> float:
        """Get wallet balance"""
        wallet = self.get_wallet(wallet_id)
        return wallet.balance
    
    # Payment Processing
    def create_transfer(
        self,
        from_wallet_id: str,
        to_wallet_id: str,
        amount: float,
        currency: str
    ) -> Transfer:
        """Create a transfer between wallets"""
        payload = {
            "from_wallet_id": from_wallet_id,
            "to_wallet_id": to_wallet_id,
            "amount": amount,
            "currency": currency
        }
        response = self.client.post(f"{self.base_url}/transfers", json=payload)
        data = self._handle_response(response)
        return Transfer(**data)
    
    def get_transfer(self, transfer_id: str) -> Transfer:
        """Get transfer details"""
        response = self.client.get(f"{self.base_url}/transfers/{transfer_id}")
        data = self._handle_response(response)
        return Transfer(**data)
    
    def on_ramp(
        self,
        customer_id: str,
        wallet_id: str,
        amount: float,
        currency: str,
        payment_method: str
    ) -> Dict[str, Any]:
        """On-ramp fiat to crypto"""
        payload = {
            "customer_id": customer_id,
            "wallet_id": wallet_id,
            "amount": amount,
            "currency": currency,
            "payment_method": payment_method
        }
        response = self.client.post(f"{self.base_url}/on-ramp", json=payload)
        return self._handle_response(response)
    
    def off_ramp(
        self,
        customer_id: str,
        wallet_id: str,
        amount: float,
        currency: str,
        destination_account: str
    ) -> Dict[str, Any]:
        """Off-ramp crypto to fiat"""
        payload = {
            "customer_id": customer_id,
            "wallet_id": wallet_id,
            "amount": amount,
            "currency": currency,
            "destination_account": destination_account
        }
        response = self.client.post(f"{self.base_url}/off-ramp", json=payload)
        return self._handle_response(response)
    
    # Card Management
    def issue_card(
        self,
        customer_id: str,
        card_type: str = "virtual",
        spending_limit: Optional[float] = None
    ) -> Card:
        """Issue a new card"""
        payload = {
            "customer_id": customer_id,
            "card_type": card_type,
            "spending_limit": spending_limit
        }
        response = self.client.post(f"{self.base_url}/cards", json=payload)
        data = self._handle_response(response)
        return Card(**data)
    
    def get_card(self, card_id: str) -> Card:
        """Get card details"""
        response = self.client.get(f"{self.base_url}/cards/{card_id}")
        data = self._handle_response(response)
        return Card(**data)
    
    def freeze_card(self, card_id: str) -> bool:
        """Freeze a card"""
        response = self.client.post(f"{self.base_url}/cards/{card_id}/freeze")
        self._handle_response(response)
        return True
    
    def cancel_card(self, card_id: str) -> bool:
        """Cancel a card"""
        response = self.client.delete(f"{self.base_url}/cards/{card_id}")
        self._handle_response(response)
        return True
    
    # External Accounts
    def add_external_account(
        self,
        customer_id: str,
        account_number: str,
        routing_number: str,
        account_type: str = "checking"
    ) -> Dict[str, Any]:
        """Add external bank account"""
        payload = {
            "customer_id": customer_id,
            "account_number": account_number,
            "routing_number": routing_number,
            "account_type": account_type
        }
        response = self.client.post(f"{self.base_url}/external-accounts", json=payload)
        return self._handle_response(response)
    
    def list_external_accounts(self, customer_id: str) -> List[Dict[str, Any]]:
        """List external accounts"""
        params = {"customer_id": customer_id}
        response = self.client.get(f"{self.base_url}/external-accounts", params=params)
        data = self._handle_response(response)
        return data.get("accounts", [])
    
    # Webhook Management
    def register_webhook(self, url: str, events: List[str]) -> Dict[str, Any]:
        """Register a webhook endpoint"""
        payload = {"url": url, "events": events}
        response = self.client.post(f"{self.base_url}/webhooks", json=payload)
        return self._handle_response(response)
    
    def verify_webhook_signature(
        self,
        payload: str,
        signature: str,
        secret: str
    ) -> bool:
        """Verify webhook signature"""
        import hmac
        import hashlib
        expected = hmac.new(
            secret.encode(),
            payload.encode(),
            hashlib.sha256
        ).hexdigest()
        return hmac.compare_digest(expected, signature)
    
    def close(self):
        """Close the HTTP client"""
        self.client.close()
    
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()


# Convenience function
def create_bridge_client(
    api_key: Optional[str] = None,
    environment: str = "sandbox"
) -> BridgeSDK:
    """Create a Bridge SDK client instance"""
    api_key = api_key or os.getenv("BRIDGE_API_KEY")
    if not api_key:
        raise ValueError("Bridge API key is required")
    
    env = BridgeEnvironment.SANDBOX if environment == "sandbox" else BridgeEnvironment.PRODUCTION
    return BridgeSDK(api_key=api_key, environment=env)

File: bridge-integration/sdk/typescript/src/bridge-sdk.ts
/**
 * Bridge API SDK for TypeScript
 * Complete integration with Bridge payment platform
 */
import axios, { AxiosInstance, AxiosError } from 'axios';
import type {
  BridgeConfig,
  BridgeEnvironment,
  Customer,
  Wallet,
  Transfer,
  Card,
  ExternalAccount,
  WalletType,
  APIResponse
} from './types';

export class BridgeAPIError extends Error {
  constructor(
    message: string,
    public statusCode?: number,
    public errorCode?: string
  ) {
    super(message);
    this.name = 'BridgeAPIError';
  }
}

export class BridgeSDK {
  private client: AxiosInstance;
  private apiKey: string;
  private environment: BridgeEnvironment;

  constructor(config: BridgeConfig) {
    this.apiKey = config.apiKey;
    this.environment = config.environment || 'sandbox';
    
    const baseURL = this.getBaseURL();
    
    this.client = axios.create({
      baseURL,
      timeout: config.timeout || 30000,
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
        'User-Agent': 'BridgeSDK-TypeScript/1.0'
      }
    });

    this.setupInterceptors();
  }

  private getBaseURL(): string {
    return this.environment === 'sandbox'
      ? 'https://api.sandbox.bridge.xyz/v1'
      : 'https://api.bridge.xyz/v1';
  }

  private setupInterceptors(): void {
    this.client.interceptors.response.use(
      response => response,
      (error: AxiosError) => {
        if (error.response) {
          const errorData = error.response.data as any;
          throw new BridgeAPIError(
            errorData?.message || 'Bridge API error',
            error.response.status,
            errorData?.code
          );
        }
        throw error;
      }
    );
  }

  // Customer Management
  async createCustomer(
    email: string,
    firstName?: string,
    lastName?: string,
    metadata?: Record<string, any>
  ): Promise<Customer> {
    const response = await this.client.post<Customer>('/customers', {
      email,
      first_name: firstName,
      last_name: lastName,
      metadata: metadata || {}
    });
    return response.data;
  }

  async getCustomer(customerId: string): Promise<Customer> {
    const response = await this.client.get<Customer>(`/customers/${customerId}`);
    return response.data;
  }

  async listCustomers(limit: number = 100, offset: number = 0): Promise<Customer[]> {
    const response = await this.client.get<{ customers: Customer[] }>('/customers', {
      params: { limit, offset }
    });
    return response.data.customers;
  }

  async deleteCustomer(customerId: string): Promise<boolean> {
    await this.client.delete(`/customers/${customerId}`);
    return true;
  }

  async startKYC(customerId: string): Promise<any> {
    const response = await this.client.post(`/customers/${customerId}/kyc`);
    return response.data;
  }

  // Wallet Management
  async createWallet(
    customerId: string,
    currency: string,
    walletType: WalletType = 'crypto'
  ): Promise<Wallet> {
    const response = await this.client.post<Wallet>('/wallets', {
      customer_id: customerId,
      currency,
      wallet_type: walletType
    });
    return response.data;
  }

  async getWallet(walletId: string): Promise<Wallet> {
    const response = await this.client.get<Wallet>(`/wallets/${walletId}`);
    return response.data;
  }

  async listWallets(customerId: string): Promise<Wallet[]> {
    const response = await this.client.get<{ wallets: Wallet[] }>('/wallets', {
      params: { customer_id: customerId }
    });
    return response.data.wallets;
  }

  async getWalletBalance(walletId: string): Promise<number> {
    const wallet = await this.getWallet(walletId);
    return wallet.balance;
  }

  // Payment Processing
  async createTransfer(
    fromWalletId: string,
    toWalletId: string,
    amount: number,
    currency: string
  ): Promise<Transfer> {
    const response = await this.client.post<Transfer>('/transfers', {
      from_wallet_id: fromWalletId,
      to_wallet_id: toWalletId,
      amount,
      currency
    });
    return response.data;
  }

  async getTransfer(transferId: string): Promise<Transfer> {
    const response = await this.client.get<Transfer>(`/transfers/${transferId}`);
    return response.data;
  }

  async onRamp(
    customerId: string,
    walletId: string,
    amount: number,
    currency: string,
    paymentMethod: string
  ): Promise<any> {
    const response = await this.client.post('/on-ramp', {
      customer_id: customerId,
      wallet_id: walletId,
      amount,
      currency,
      payment_method: paymentMethod
    });
    return response.data;
  }

  async offRamp(
    customerId: string,
    walletId: string,
    amount: number,
    currency: string,
    destinationAccount: string
  ): Promise<any> {
    const response = await this.client.post('/off-ramp', {
      customer_id: customerId,
      wallet_id: walletId,
      amount,
      currency,
      destination_account: destinationAccount
    });
    return response.data;
  }

  // Card Management
  async issueCard(
    customerId: string,
    cardType: string = 'virtual',
    spendingLimit?: number
  ): Promise<Card> {
    const response = await this.client.post<Card>('/cards', {
      customer_id: customerId,
      card_type: cardType,
      spending_limit: spendingLimit
    });
    return response.data;
  }

  async getCard(cardId: string): Promise<Card> {
    const response = await this.client.get<Card>(`/cards/${cardId}`);
    return response.data;
  }

  async freezeCard(cardId: string): Promise<boolean> {
    await this.client.post(`/cards/${cardId}/freeze`);
    return true;
  }

  async cancelCard(cardId: string): Promise<boolean> {
    await this.client.delete(`/cards/${cardId}`);
    return true;
  }

  // External Accounts
  async addExternalAccount(
    customerId: string,
    accountNumber: string,
    routingNumber: string,
    accountType: string = 'checking'
  ): Promise<ExternalAccount> {
    const response = await this.client.post<ExternalAccount>('/external-accounts', {
      customer_id: customerId,
      account_number: accountNumber,
      routing_number: routingNumber,
      account_type: accountType
    });
    return response.data;
  }

  async listExternalAccounts(customerId: string): Promise<ExternalAccount[]> {
    const response = await this.client.get<{ accounts: ExternalAccount[] }>(
      '/external-accounts',
      { params: { customer_id: customerId } }
    );
    return response.data.accounts;
  }

  // Webhook Management
  async registerWebhook(url: string, events: string[]): Promise<any> {
    const response = await this.client.post('/webhooks', { url, events });
    return response.data;
  }

  async verifyWebhookSignature(
    payload: string,
    signature: string,
    secret: string
  ): Promise<boolean> {
    const crypto = await import('crypto');
    const expected = crypto
      .createHmac('sha256', secret)
      .update(payload)
      .digest('hex');
    return crypto.timingSafeEqual(
      Buffer.from(expected),
      Buffer.from(signature)
    );
  }
}

// Convenience factory function
export function createBridgeClient(
  apiKey?: string,
  environment: BridgeEnvironment = 'sandbox'
): BridgeSDK {
  const key = apiKey || process.env.BRIDGE_API_KEY;
  if (!key) {
    throw new Error('Bridge API key is required');
  }
  return new BridgeSDK({ apiKey: key, environment });
}
